core/src/main/scala/org/apache/spark/SparkEnv.scala

```diff
@@ -355,6 +355,11 @@ object SparkEnv extends Logging {
       None
     }
 
+    // Set the application attemptId in the ExternalShuffleClient is applicable.
+    // If there is no attemptId assigned, set the attemptId to -1.
+    externalShuffleClient.foreach(
+      shuffleClient => shuffleClient.setAppAttemptId(conf.getInt(config.APP_ATTEMPT_ID.key, -1)))
```

Comment:
Shall we also set appAttemptId after the executor's SparkEnv is created? Just like what driver does. e.g., we can set after:
https://github.com/apache/spark/blob/67421d80b8935d91b86e8cd3becb211fa2abd54f/core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala#L471-L472

Reply:
I don't think we need to change the CoarseGrainedExecutorBackend to set the attemptId. The _conf.set(APP_ATTEMPT_ID, attemptId) in SparkContext happens before the Executor starts.
Executor will fetch the Spark confs from Driver, and Driver will respond the confs. 
The [SparkProperties here is lazy val](https://github.com/apache/spark/blob/917d7dad4dcdbeac5094899fa9b7fffc67376cec/core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala#L132), so it will include all the conf changes made in Driver.
Then in Executor, after it [gets the Spark confs](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala#L451) from Driver, it will[ create the SparkEnv after that](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala#L471).
Then the SparkEnv will have the attemptId in the conf.
The test in our Yarn cluster runs fine, so Executor should have the attemptId set, otherwise, the push blocks from Executor will fail.

Reply:
I know the executor can get the right attempted. My concern is that keep setting it here is wrong in terms of the driver's view. People could be confused by setting 2 times.  Besides, setting after SparkEnv created for the executor makes the behavior be consistent.

Reply:
Make sense. Updated. Tested again within the cluster which worked fine.

Reply:
Thanks!